spec/compromise-syntax.md

```diff
@@ -0,0 +1,238 @@
+the prefix character for an argument name, literal value, or function;
+and after the function name there must be only white-space-separated options which
+start with identifier-start characters.
+
+# Syntactic white space
+
+We use white space inside placeholders and in “code mode” (outside patterns) as token separators.
+White space is a sequence of one or more of the characters TAB, LF, CR, SP, and maybe some more.
+- TODO: For the definition of white space we should consult with the Unicode Source Code Working Group.
+- TODO: Decide whether to use Unicode Pattern_White_Space or otherwise allow RLM and LRM characters.
+
+White space can also be useful for line breaking long messages, indentation, and alignment.
+However, we should not allow white space everywhere possible,
+because that just leads to confusing variations in style,
+and the creation of formatting tools to enforce certain styles.
+For example, there is no reason to allow white space between a name or function prefix and its identifier,
+around the `=` of an option, after the `{` of a placeholder, or before the `}` of a placeholder.
+
+# Pattern selection
+
+Messages need the ability to choose among variants of a pattern based on certain argument values.
+Common examples include selecting the right plural form, and variants for different person genders.
+
+There should be a single level of selection (not nested like in ICU MessageFormat).
+It needs to support multiple selectors.
+
+In this syntax, a list of N selectors is followed by a list of pairs where
+the first element of each pair is a list of N value literals and
+the second element of each pair is a pattern.
+A `_` is a wildcard value that always matches.
+The last variant must have a list of all wildcard values.
+```
+[{$count :plural offset=1 grouping=always} {$gender}]
+[1 female] {{$name} added you to her circles.}
+[1 male] {{$name} added you to his circles.}
+[1 _] {{$name} added you to their circles.}
+[_ _] {{$name} added you and {#count} others to their circles.}
+```
+
+Lists are enclosed in square brackets, reminiscent of Python lists.
+The opening `[` also distinguishes the selection syntax from a simple pattern.
+
+TODO: Decide whether to enclose each value literal in
+the same pair of delimiters as literals in placeholder (for consistency),
+or whether to make that optional.
+(The `[]` value list syntax already indicates that value literals are enclosed.)
+Some literals may require it if they contain spaces.
+The `_` should probably never be enclosed in literal delimiters.
+
+Selector syntax follows placeholder syntax,
+except that a function must be specified.
+For the purpose of selection, there are three types of functions:
+1. Select-and-format functions combine the two functionalities,
+   and the selection is informed by the formatting.
+   For example, selectors for plural variants
+   (different selectors for cardinal-number vs. ordinal-number variants)
```

Comment:
that's not formatting: that's the selector type. If I say `[{$count :plural type=ordinal}]` I expect to get keywords out like `one`, `few`, etc. or access the numeric value of `$count` for selectors such as `=2`---just like plural rules work today.

I agree that options are needed for the selector (as shown), but tend to expect that I can still format the value with a placeholder later. In fact, I might format differently several times.

Reply:
> that's not formatting: that's the selector type. If I say `[{$count :plural type=ordinal}]` I expect to get keywords out like `one`, `few`, etc. or access the numeric value of `$count` for selectors such as `=2`---just like plural rules work today.

I assumed that there would be different function names for plural/cardinal vs. plural/ordinal, like we have in ICU MessageFormat. But yes, it could be one "plural" function with an option.

> I agree that options are needed for the selector (as shown), but tend to expect that I can still format the value with a placeholder later. In fact, I might format differently several times.

That should be strongly discouraged, especially looking at plurals. Formatting differently from what the selection was based on creates a jarring mismatch. We should design this to make it easy to do the right thing. If you need different formatting in a different part of the sentence, you can pass the same value in another argument, and you can also define a named expression for it.